runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/sideinput/SideInputStateFetcher.java

```diff
@@ -0,0 +1,257 @@
+        }
+
+        if (!existingCacheEntry.get().isReady()) {
+          return sideInputCache.invalidateThenLoadNewEntry(
+              sideInputCacheKey, loadSideInputFromWindmill);
+        }
+
+        return existingCacheEntry.get();
+      }
+
+      return sideInputCache.getOrLoad(sideInputCacheKey, loadSideInputFromWindmill);
+    } catch (Exception e) {
+      LOG.error("Fetch failed: ", e);
+      throw new RuntimeException("Exception while fetching side input: ", e);
+    }
+  }
+
+  private <T, SideWindowT extends BoundedWindow> GlobalData fetchGlobalDataFromWindmill(
+      PCollectionView<T> view,
+      SideWindowT sideWindow,
+      String stateFamily,
+      Supplier<Closeable> scopedReadStateSupplier)
+      throws IOException {
+    @SuppressWarnings({
+      "deprecation", // Internal windowStrategy is required to fetch side input data from Windmill.
+      "unchecked" // Internal windowing strategy matches WindowingStrategy<?, SideWindowT>.
+    })
+    WindowingStrategy<?, SideWindowT> sideWindowStrategy =
+        (WindowingStrategy<?, SideWindowT>) view.getWindowingStrategyInternal();
+
+    Coder<SideWindowT> windowCoder = sideWindowStrategy.getWindowFn().windowCoder();
+
+    ByteStringOutputStream windowStream = new ByteStringOutputStream();
+    windowCoder.encode(sideWindow, windowStream);
+
+    @SuppressWarnings({
+      "deprecation" // PCollectionView's internal tag needed to create a GlobalDataId.
+    })
+    Windmill.GlobalDataRequest request =
+        Windmill.GlobalDataRequest.newBuilder()
+            .setDataId(
+                Windmill.GlobalDataId.newBuilder()
+                    .setTag(view.getTagInternal().getId())
+                    .setVersion(windowStream.toByteString())
+                    .build())
+            .setStateFamily(stateFamily)
+            .setExistenceWatermarkDeadline(
+                WindmillTimeUtils.harnessToWindmillTimestamp(
+                    sideWindowStrategy.getTrigger().getWatermarkThatGuaranteesFiring(sideWindow)))
+            .build();
+
+    try (Closeable ignored = scopedReadStateSupplier.get()) {
+      return server.getSideInputData(request);
+    }
+  }
+
+  private <T> SideInput<T> loadSideInputFromWindmill(
+      PCollectionView<T> view,
+      BoundedWindow sideWindow,
+      String stateFamily,
+      Supplier<Closeable> scopedReadStateSupplier)
+      throws IOException {
+    validateViewMaterialization(view);
+    GlobalData data =
+        fetchGlobalDataFromWindmill(view, sideWindow, stateFamily, scopedReadStateSupplier);
+    bytesRead += data.getSerializedSize();
+    return data.getIsReady() ? createSideInputCacheEntry(view, data) : SideInput.notReady();
+  }
+
+  private <T> void validateViewMaterialization(PCollectionView<T> view) {
+    @SuppressWarnings({
+      "deprecation" // Underlying ViewFn is needed to validate the materialization URN.
+    })
+    String materializationUrn = view.getViewFn().getMaterialization().getUrn();
```

Comment:
how about a private helper method to get urn from view so you don't have to repeat suppression and explanation a lot

Reply:
done.